### PR TITLE
Tickets/dm 30949

### PR DESCRIPTION
**Pull request by tribeiro**
* Update configuration schema to remove deprecated schema file and use module instead.
* Add hexapod sensitivity matrix to the model.
    This allows users to specify linear cross-terms between different hexapod axis.
    For instance, if one wants to apply a tip/tilt correction when doing x/y translation, it is possible to specify the cross term in the sensitivity matrix.
* Add feature that allow users to specify the valid range for LUTs.
    Any data beyond the limits return the value of the LUT in the limit.
    For instance, if minimum valid elevation for hexapod LUT is 30 and a correction is requested for elevation 20, it returns the value in elevation 30.

### CODE REVIEW

File: python/lsst/ts/ataos/config_schema.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    items:
+      type: number
+    default: [0.]
+  hexapod_v:
+    description: List of polynomial coefficients for hexapod v-correction equation.
+    type: array
+    items:
+      type: number
+    default: [0.]
+  hexapod_sensitivity_matrix:
```
Comment (patrickingraham) [inline python/lsst/ts/ataos/config_schema.py:116]: So I think there is a bit of an issue here since we have the matrix in the script and the matrix here. I think the answer is obviously to not hard-code the matrix in the script, but what is the clean way to do that?
Does the script pull it out of the configuration that gets published by the ATAOS (assuming it's actually getting published?)

File: python/lsst/ts/ataos/config_schema.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
+# This file is part of ts_scheduler.
```
Comment (patrickingraham) [inline python/lsst/ts/ataos/config_schema.py:1]: uhh, don't think so :)

File: python/lsst/ts/ataos/config_schema.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+  m1:
+    description: List of polynomial coefficients for M1 correction equation.
+    type: array
+    items:
+      type: number
+    default: [0.]
+  m1_lut_elevation_limits:
+    description: >-
+        Elevation limits for the validity of the M1 pressure LUT. Any values
+        outside this range you be assigned the value at this position.
```
Comment (patrickingraham) [inline python/lsst/ts/ataos/config_schema.py:46]: ...outside this range will be assigned....

File: python/lsst/ts/ataos/config_schema.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+  m2:
+    description: List of polynomial coefficients for M2 correction equation.
+    type: array
+    items:
+      type: number
+    default: [0.]
+  m2_lut_elevation_limits:
+    description: >-
+        Elevation limits for the validity of the M2 pressure LUT. Any values
+        outside this range you be assigned the value at this position.
```
Comment (patrickingraham) [inline python/lsst/ts/ataos/config_schema.py:64]: will be assigned....

File: python/lsst/ts/ataos/config_schema.py (hunk 3 above)
Comment (patrickingraham) [reply python/lsst/ts/ataos/config_schema.py:46]: So this is a little dangerous actually as it can lead to mirror lifting. For example, if you put the lower limit at 20 degrees, then when you go to 10 the pressure will be set for 20, which might cause a lift. Maybe it doesn't matter?
Another way to do this would be to just make the function linear to zero?

File: python/lsst/ts/ataos/config_schema.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    default: [0.]
+  hexapod_v:
+    description: List of polynomial coefficients for hexapod v-correction equation.
+    type: array
+    items:
+      type: number
+    default: [0.]
+  hexapod_sensitivity_matrix:
+    description: >-
+        A matrix to map the cross terms dependencies between hexapod axis.
```
Comment (patrickingraham) [inline python/lsst/ts/ataos/config_schema.py:118]: is this written up in a technote somewhere we can reference?

File: python/lsst/ts/ataos/model.py [hunk 6]
```diff
@@ -44,6 +44,21 @@ def __init__(self, log):
             "v": 0.0,
         }
 
+        # Hexapod sensitivity matrix, this is to account for cross terms
+        # between the axis. By default, there are no cross terms. This can be
+        # used, for instance, to add automatic tip-tilt correction for a given
+        # hexapod translation (x/y).
+        self._hexapod_sensitivity_matrix = np.array(
```
Comment (patrickingraham) [inline python/lsst/ts/ataos/model.py:51]: this comes from the config, no? why is it defined here?

File: python/lsst/ts/ataos/model.py [hunk 7]
```diff
@@ -53,6 +68,10 @@ def __init__(self, log):
         self.poly_v = np.poly1d(self.config["hexapod_v"])
         self.poly_chromatic = np.poly1d(self.config["chromatic_dependence"])
 
+        self.m1_lut_elevation_limits = [0.0, 90.0]
```
Comment (patrickingraham) [inline python/lsst/ts/ataos/model.py:71]: same deal, this is in the config, same with the lines below

File: python/lsst/ts/ataos/model.py [hunk 8]
```diff
@@ -270,3 +326,25 @@ def chromatic_dependence(self):
     def chromatic_dependence(self, val):
         self.config["chromatic_dependence"] = val
         self.poly_chromatic = np.poly1d(val)
+
+    @staticmethod
+    def get_lut_elevation(elevation, limits):
+        """Return an elevation inside the rande.
```
Comment (patrickingraham) [inline python/lsst/ts/ataos/model.py:332]: inside the rande?   I'm guessing you mean range?

File: doc/version_history.rst [hunk 9]
```diff
@@ -6,6 +6,19 @@
 Version History
 ###############
 
+v1.8.0
+------
+
+Changes:
+
+  * Update configuration schema to remove schema file and use module instead.
```
Comment (patrickingraham) [inline doc/version_history.rst:14]: The configs are all backwards compatible, yes? No new version is required?

File: python/lsst/ts/ataos/model.py (hunk 6 above)
Comment (patrickingraham) [reply python/lsst/ts/ataos/model.py:51]: Also, I think we should define a test for this, where the cross-terms are non-zero of course, no?

**Review by patrickingraham — CHANGES_REQUESTED** [review]

File: python/lsst/ts/ataos/model.py (hunk 6 above)
Comment (tribeiro) [reply python/lsst/ts/ataos/model.py:51]: So, this is supposed to be an independent class that can run and test without the CSC. The configuration belongs to the CSC business logic, which this class has no knowledge about. 

I can't remove these default values. What I can do is remove the defaults from the configuration and make the CSC not set them if the user does not provide some values. Nevertheless, it also means users won't have anywhere to look for an example of the configuration. Which maybe could be done in documentation.

File: python/lsst/ts/ataos/config_schema.py (hunk 1 above)
Comment (tribeiro) [reply python/lsst/ts/ataos/config_schema.py:116]: When you say "script" are you referring to the `Model` class? The class itself has no knowledge of the CSC so it doesn't do any publishing. The CSC can most certainly publish these values, though it is not doing so yet. Part of the issue of publishing this CSC configuration is that the number of polynomial terms are unrestricted. So we would have to publish them as comma separated strings...

File: python/lsst/ts/ataos/config_schema.py (hunk 3 above)
Comment (tribeiro) [reply python/lsst/ts/ataos/config_schema.py:46]: I think you are right that it might be a bit prone to lifting the mirror, especially at these lower elevations. It could as well switch to a linear fit bellow the limit. Probably not going to zero though, maybe the user provide an inclination and we computer the zero-point from the value in the limit?

File: python/lsst/ts/ataos/config_schema.py (hunk 5 above)
Comment (tribeiro) [reply python/lsst/ts/ataos/config_schema.py:118]: No.... Could probably add it to the package documentation.

File: python/lsst/ts/ataos/model.py (hunk 7 above)
Comment (tribeiro) [reply python/lsst/ts/ataos/model.py:71]: Same argument... :)

File: doc/version_history.rst (hunk 9 above)
Comment (tribeiro) [reply doc/version_history.rst:14]: correct

**Review by patrickingraham — APPROVED** [review]
